### ISSUE TITLE
nchan_publisher_upstream_request with chunked responses publish empty messages

### ISSUE THREAD

**Issue by the-noob** [slact/nchan#281]
Probably I'm doing something wrong but can't figure out what.
Server config:
```nginx
    location ~ /pub/(\w+)$ {
        nchan_channel_id $1;
        nchan_publisher;
        nchan_publisher_upstream_request /upstream_pub;
        nchan_websocket_ping_interval 15;
    }

    location ~ /sub/(\w+)$ {
      nchan_subscriber;
      nchan_channel_id $1;
      nchan_websocket_ping_interval 15;
    }

    location = /upstream_pub {
        proxy_pass http://127.0.0.1/chat.php;
        proxy_set_header X-Publisher-Type $nchan_publisher_type;
        proxy_set_header X-Prev-Message-Id $nchan_prev_message_id;
        proxy_set_header X-Channel-Id $nchan_channel_id;
        proxy_set_header X-Original-URI $request_uri;
    }
```
The server side seems to be working, when doing a POST I get a proper response and things happen in my PHP script.

    curl --request POST --data "test message" -H "Accept: text/json" http://127.0.0.2:80/pub/chat
    {"messages": 5, "requested": 11, "subscribers": 1, "last_message_id": "1479388144:0" }

For the client I've tried using direct WebSocket:
```js
    sub = new WebSocket("ws://example.com/sub/chat", "ws+meta.nchan");

    sub.onmessage = function(e) {
      console.log(['protocol',sub.protocol]);
      console.log(['message',e.data]); // e.data is empty
    }
```
And tried with `nchan/NchanSubscriber.js`
```js
    var sub = new NchanSubscriber("http://example.com/sub/chat", ['longpoll', 'websocket']);
    //tried with ws+meta.nchan://example.com/sub/chat too
    sub.on("message", function(message, message_metadata) {

     console.log(['msg', message, message_metadata]); // message is an empty string
    });
```

**Comment by the-noob** [comment]
I've added a SO question if anyone needs some rep :) http://stackoverflow.com/questions/40660164/

**Comment by slact** [comment]
What does your `chat.php` script look like? I would bet that it returns an empty `200 OK` response. You should instead return code `304 Not Modified`.

From the documentation: 

> The upstream response code determine how publishing will proceed. A `200 OK `will publish the message from the upstream response's body. A `304 Not Modified` will publish the message as it was received from the publisher. A `204 No Content` will result in the message not being published.

https://nchan.slact.net/#nchan_publisher_upstream_request

**Comment by the-noob** [comment]
Almost there!

Returning `304 Not Modified` pushes a `message=...` message but if I echo something in the script it returns an empty string. 

    <?php
    echo 'something';

**Comment by slact** [comment]
Please do a `curl -v http://127.0.0.1/chat.php` and show me its output.

**Comment by slact** [comment]
Actually, two things:
`curl -v http://127.0.0.1/chat.php`
`curl --request POST --data "test message" -H "Accept: text/json"  "http://127.0.0.1/chat.php"`

**Comment by the-noob** [comment]
curl -v http://127.0.0.1/chat.php
    *   Trying 127.0.0.1...
    * Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
    > GET /chat.php HTTP/1.1
    > Host: 127.0.0.1
    > User-Agent: curl/7.47.1
    > Accept: */*
    > 
    < HTTP/1.1 200 OK
    < Server: nginx/1.10.2
    < Date: Thu, 17 Nov 2016 20:15:40 GMT
    < Content-Type: text/html; charset=UTF-8
    < Transfer-Encoding: chunked
    < Connection: keep-alive
    < X-Powered-By: PHP/7.0.11
    < 
    * Connection #0 to host 127.0.0.1 left intact


    curl -v --request POST --data "test message" -H "Accept: text/json" "http://127.0.0.1/chat.php"
    Note: Unnecessary use of -X or --request, POST is already inferred.
    *   Trying 127.0.0.1...
    * Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
    > POST /chat.php HTTP/1.1
    > Host: 127.0.0.1
    > User-Agent: curl/7.47.1
    > Accept: text/json
    > Content-Length: 12
    > Content-Type: application/x-www-form-urlencoded
    > 
    * upload completely sent off: 12 out of 12 bytes
    < HTTP/1.1 200 OK
    < Server: nginx/1.10.2
    < Date: Thu, 17 Nov 2016 20:16:18 GMT
    < Content-Type: text/html; charset=UTF-8
    < Transfer-Encoding: chunked
    < Connection: keep-alive
    < X-Powered-By: PHP/7.0.11
    < 
    * Connection #0 to host 127.0.0.1 left intact

**Comment by slact** [comment]
It looks like your php script isn't outputting any data...

**Comment by the-noob** [comment]
sorry, forgot to revert some changes

    cat /usr/share/nginx/html/chat.php
    <?php

    require __DIR__ . '/bootstrap.php';

    if ($_POST) {
       $messages[] = $_POST;
       echo current($_POST) . "\n";
       file_put_contents($storage, json_encode($messages));
    } else {
      echo "hello from chat.php\n";
    }


    curl -v http://127.0.0.1/chat.php
    *   Trying 127.0.0.1...
    * Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
    > GET /chat.php HTTP/1.1
    > Host: 127.0.0.1
    > User-Agent: curl/7.47.1
    > Accept: */*
    > 
    < HTTP/1.1 200 OK
    < Server: nginx/1.10.2
    < Date: Thu, 17 Nov 2016 20:23:41 GMT
    < Content-Type: text/html; charset=UTF-8
    < Transfer-Encoding: chunked
    < Connection: keep-alive
    < X-Powered-By: PHP/7.0.11
    < 
    hello from chat.php
    * Connection #0 to host 127.0.0.1 left intact

curl -v --request POST --data "test message" -H "Accept: text/json" "http://127.0.0.1/chat.php"
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> POST /chat.php HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.47.1
> Accept: text/json
> Content-Length: 12
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 12 out of 12 bytes
< HTTP/1.1 200 OK
< Server: nginx/1.10.2
< Date: Thu, 17 Nov 2016 20:26:29 GMT
< Content-Type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< X-Powered-By: PHP/7.0.11
< 

* Connection #0 to host 127.0.0.1 left intact

**Comment by the-noob** [comment]
I've just simplified the script making a simple (different ) echo if POST, still no messages in websockets 

    curl -v --request POST --data "test message" -H "Accept: text/json" "http://127.0.0.1/chat.php"
    Note: Unnecessary use of -X or --request, POST is already inferred.
    *   Trying 127.0.0.1...
    * Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
    > POST /chat.php HTTP/1.1
    > Host: 127.0.0.1
    > User-Agent: curl/7.47.1
    > Accept: text/json
    > Content-Length: 12
    > Content-Type: application/x-www-form-urlencoded
    > 
    * upload completely sent off: 12 out of 12 bytes
    < HTTP/1.1 200 OK
    < Server: nginx/1.10.2
    < Date: Thu, 17 Nov 2016 20:34:34 GMT
    < Content-Type: text/html; charset=UTF-8
    < Transfer-Encoding: chunked
    < Connection: keep-alive
    < X-Powered-By: PHP/7.0.11
    < 
    hello from post
    * Connection #0 to host 127.0.0.1 left intact

**Comment by slact** [comment]
Any chance you could show me the server where this is happening? You can email me at nchan@slact.net

**Comment by slact** [comment]
The issue is with `nchan_publisher_upstream_request` incorrectly handling `Transfer-Encoding: chunked` responses.

**Comment by the-noob** [comment]
Just for the record, if someone needs an urgent fix, `header('Content-length: '. strlen($output));` will fix things. (assuming you're buffering the output in a variable and it is ASCII)

**Comment by slact** [comment]
fixed in 4d100ba533d218. Please re-test to verify when you are able.

**Comment by slact** [comment]
fixed even more in 99adffc8b1d59811feaf3f5dd19
